### PR TITLE
JIT: improve profile update for loop inversion

### LINKED ISSUE #84319: JIT: loop inversion doing some questionable profile maintenance
From the `PerfLabTests.CastingPerf:CheckArrayIsInterfaceNo` benchmark, running with TieredPGO, during an OSR recompile:

Note how BB04's profile data gets trashed
```
****** START compiling PerfLabTests.CastingPerf:CheckArrayIsInterfaceNo():bool:this (MethodHash=a46d9cd6)
Generating code for Windows x64
OPTIONS: Tier-1 compilation
OPTIONS: OSR variant with entry point 0x18
...

*************** Starting PHASE Invert loops
-----------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight       IBC  lp [IL range]     [jump]      [EH region]         [flags]
-----------------------------------------------------------------------------------------------------------------------------------------
BB01 [0004]  1                             1      48701    [???..???)-> BB04 (always)                     i internal IBC 
BB02 [0000]  0                             0.00       1    [000..006)-> BB04 (always)                     i IBC 
BB03 [0001]  1       BB04                100.00 4870143    [006..018)                                     i hascall bwd bwd-target IBC 
BB04 [0002]  3       BB01,BB02,BB03      100    4870144    [018..020)-> BB03 ( cond )                     i bwd bwd-src IBC  osr-entry
BB05 [0003]  1       BB04                  0          0    [020..022)        (return)                     i rare IBC 
-----------------------------------------------------------------------------------------------------------------------------------------

***************  Exception Handling table is empty
=============== No blocks renumbered!

Duplication of loop condition [000010] is performed, because the cost of duplication (17) is less or equal than 136,
   loopIterations = 4870143.000, optInvertTotalInfo.sharedStaticHelperCount >= 0, validProfileWeights = true
New Basic Block BB06 [0005] created.
Setting edge weights for BB02 -> BB06 to [0 .. 1]
Setting edge weights for BB06 -> BB03 to [0 .. 1]
Setting edge weights for BB06 -> BB05 to [0 .. 0]
Redirecting non-loop BB01 -> BB04 to BB01 -> BB06
Setting edge weights for BB01 -> BB06 to [0 .. 1]
Reducing profile weight of BB04 from 4870144 to 4870143
Setting weight of BB04 -> BB03 to 4870142 (iterate loop)
Setting weight of BB04 -> BB05 to 0.9999998 (exit loop)
Setting edge weights for BB04 -> BB03 to [4870142 .. 4870142]
Setting edge weights for BB04 -> BB05 to [0.9999998 .. 0.9999998]
Profile weight of BB02 remains unchanged at 1
Setting weight of BB06 -> BB03 to 0.9999998 (enter loop)
Setting weight of BB06 -> BB05 to 2.053327e-07 (avoid loop)
Setting edge weights for BB06 -> BB03 to [0.9999998 .. 0.9999998]
Setting edge weights for BB06 -> BB05 to [2.053327e-07 .. 2.053327e-07]

Duplicated loop exit block at BB06 for loop (BB03 - BB04)
Estimated code size expansion is 17

------------ BB06 [???..???) -> BB05 (cond), preds={BB01,BB02} succs={BB03,BB05}

***** BB06
STMT00006 ( 0x018[E-] ... ??? )
     ( 11, 17) [000028] ----G------                         *  JTRUE     void  
     (  9, 15) [000029] J---G--N---                         \--*  GE        int   
     (  3,  2) [000030] -----------                            +--*  LCL_VAR   int    V02 loc1         
     (  5, 12) [000031] n---G------                            \--*  IND       int   
     (  3, 10) [000032] H----------                               \--*  CNS_INT(h) long   0x7ffefe7b4cb0 static Fseq[InnerIterationCount]

------------ BB04 [018..020) -> BB03 (cond), preds={BB03} succs={BB05,BB03}

***** BB04
STMT00002 ( 0x018[E-] ... 0x01E )
     ( 11, 17) [000010] ----G------                         *  JTRUE     void  
     (  9, 15) [000009] J---G--N---                         \--*  LT        int   
     (  3,  2) [000006] -----------                            +--*  LCL_VAR   int    V02 loc1         
     (  5, 12) [000008] n---G------                            \--*  IND       int   
     (  3, 10) [000007] H----------                               \--*  CNS_INT(h) long   0x7ffefe7b4cb0 static Fseq[InnerIterationCount]

*************** Finishing PHASE Invert loops
Trees after Invert loops

-----------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight       IBC  lp [IL range]     [jump]      [EH region]         [flags]
-----------------------------------------------------------------------------------------------------------------------------------------
BB01 [0004]  1                             1      48701    [???..???)-> BB06 (always)                     i internal IBC 
BB02 [0000]  0                             0.00       1    [000..006)                                     i IBC 
BB06 [0005]  2       BB01,BB02             0.00       1    [???..???)-> BB05 ( cond )                     internal IBC 
BB03 [0001]  2       BB04,BB06           100.00 4870143    [006..018)                                     i hascall bwd bwd-target IBC 
BB04 [0002]  1       BB03                  0.00       1    [018..020)-> BB03 ( cond )                     i bwd bwd-src IBC  osr-entry
BB05 [0003]  2       BB04,BB06             0          0    [020..022)        (return)                     i rare IBC 
-----------------------------------------------------------------------------------------------------------------------------------------

### PR DESCRIPTION
**Pull request by AndyAyersMS**
If the loop test block has multiple predecessors we will not do proper profile updates. This can lead to downstream problems with block layout (say leaving a cold block in a loop).

Fix by changing how we compute the amount of profile that should remain in the test block.

Fixes #84319.

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=251604&view=ms.vss-build-web.run-extensions-tab)

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib 

I've had these changes sitting around for awhile but didn't have a case where it mattered. Now I do:  https://github.com/dotnet/runtime/issues/84264#issuecomment-1517964221

Small number of diffs in SPMI PGO runs (though oddly not in the benchmark above -- likely because of how we do that collection).

File: src/coreclr/jit/optimizer.cpp [hunk 1]
```diff
@@ -5189,8 +5224,8 @@ bool Compiler::optInvertWhileLoop(BasicBlock* block)
         // cases of stress modes with inconsistent weights.
         //
         JITDUMP("Reducing profile weight of " FMT_BB " from " FMT_WT " to " FMT_WT "\n", bTest->bbNum, weightTest,
-                weightNext);
-        bTest->inheritWeight(block->bbNext);
+                weightTop);
+        bTest->inheritWeight(bTop);
```
Comment (AndyAyersMS) [inline src/coreclr/jit/optimizer.cpp:5228]: This is the actual fix -- the rest is just renaming and protecting against inconsistent profile data.

**Review by BruceForstall — APPROVED** [review]

**Comment by lewing** [comment]
please don't try to clear the `license/cla` check here, it is being looked into by the responsible team

**Comment by lewing** [comment]
investigation is done

**Comment by AndyAyersMS** [comment]
Looking like this may have had mixed impact on benchmarks (just have arm64 autofiled so far)

Improvements:
* https://github.com/dotnet/perf-autofiling-issues/issues/17163
* https://github.com/dotnet/perf-autofiling-issues/issues/17112
* https://github.com/dotnet/perf-autofiling-issues/issues/17095

Regressions:
* https://github.com/dotnet/perf-autofiling-issues/issues/17156
* https://github.com/dotnet/perf-autofiling-issues/issues/17107
* https://github.com/dotnet/perf-autofiling-issues/issues/17090

**Pull request merged**